linera-execution/src/lib.rs

```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
 use std::{io, path::Path, str::FromStr, sync::Arc};
 use thiserror::Error;
 
+/// substract an amount to a balence and report an error if that is impossible
+pub fn sub_assign_fees(balance: &mut Amount, fees: Amount) -> Result<(), PricingError> {
+    Ok(balance
+        .try_sub_assign(fees)
+        .map_err(|_| ArithmeticError::Underflow)?)
+}
+
+/// The entries of the runtime related to fuel and storage
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeLimits {
+    /// The maximum size of read allowed per block
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed per block
+    pub maximum_bytes_written: u64,
+}
+
+impl Default for RuntimeLimits {
+    fn default() -> Self {
+        RuntimeLimits {
+            maximum_bytes_read: u64::MAX,
+            maximum_bytes_written: u64::MAX,
+        }
+    }
+}
+
+/// The entries of the runtime related to fuel and storage
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeCounts {
+    /// The remaining fuel available
+    pub remaining_fuel: u64,
+    /// The number of read operations
+    pub num_reads: u64,
+    /// The bytes that have been read
+    pub bytes_read: u64,
+    /// The bytes that have been written
+    pub bytes_written: u64,
```

Comment:
It would be nicer if we treated all categories the same: Either we subtract from the remaining allowance, or we add to the usage so far.

I think we'll want to get to a point where execution stops as soon as it's clear that the balance would be insufficient to pay for it, so maybe the former is better.

Or should we rather track the remaining _balance_, and make the pricing available everywhere? That seems to be the only way to stop execution as soon as the _sum_ of all used resources becomes too high.

Reply:
Yes, it would be nice, but I try to have a realistic solution here.

I tried indeed to have the execution ending as soon as possible, as I hope is clear. But the `remaining_fuel` is managed as a `u64` so I did not see a simple way to get it closer to the heart of the matter. It is done after the runtime execution.

And yes, that involves the `pricing` information being available everywhere. Since there are specific costs associated with execution and storage and each and everyone can trigger early termination, there is no way around it.

Reply:
But knowing the maximum value for fuel, bytes_written, num_reads, and bytes_read does not mean that we know how much should occur. Because if every maximum values is attained then it does not work.

Reply:
Yes, you could still cause four times as much resource usage as your balance allows.
But that's still better than 100 or 1000 times, which is entirely possible if we only go by the fixed maximum values.

Reply:
Ok, if that is what you want, then so be it.

Reply:
So, this has been implemented.